Multiple and Single Node self hosting docs (#916)
medic/cht-docs: content/en/core/overview/self-hosting-multiple-nodes.md

> ---

Reviewer: Please move this to beside the other hosting docs here: https://docs.communityhealthtoolkit.org/apps/guides/hosting/4.x/

Reviewer 2: fixed in 1f7bd06

---

### What is a clustered setup?

In a clustered CHT setup, there are multiple couchdb nodes responding to users. This is essential to horizontally scaling your CHT instance. This is one of the capabilities added to the CHT post 4.x release. In this guide, we are going to be setting up a 3 node couchdb cluster. The way the CHT is setup in this version, we require all three couchdb nodes to be up and running. Our healthcheck module determines the health of the couchdb nodes and turns off the CHT if any single node is not functional.

### What are the four nodes used for?
* Node 1 - CHT-core - runs the core functionality of the CHT like API, sentinel
* Node 2, 3 & 4 - Couchdb Nodes (A 3 node couchdb cluster)

### Networking Setup

Make sure the following ports are open for the nodes:
* `7946 TCP` - For docker communication amongst nodes
* `7946 UDP` - For docker communication amongst nodes
* `2377 TCP` - Docker cluster management communication
* `4789 UDP` - Docker overlay network traffic
* `ICMP` - For ping

**Security Note**: As a security measure, be sure to restrict the IP addresses of the four nodes only to be able to connect to these ports.

> 

Reviewer: Consider adding the setup docker instructions here too.

Reviewer 2: fixed in d929c89

Reviewer 2: oop! This was already there from another commit:

> Provision four Ubuntu servers (22.04 as of this writing) that meet our [hosting requirements](http://localhost:1313/apps/guides/hosting/requirements/) including installing Docker and Docker on all of them.

reverted in dc9f9da

## Setup Docker Swarm and Create an Overlay Network

> ### Node 1 - Main node

Reviewer: Let's just call this the CHT Core node?

```suggestion
### CHT Core node
```

1. Initialize swarm mode.

`sudo docker swarm init`

You should get the output:

```
Swarm initialized: current node (some-long-id) is now a manager.
To add a worker to this swarm, run the following command:

    docker swarm join --token <very-long-token-value> <main-server-private-ip>:2377

To add a manager to this swarm, run 'docker swarm join-token manager' and follow the instructions.

```

2. Create overlay network

`sudo docker network create --driver=overlay --attachable cht-overlay`

> ### Node2, Node3 & Node4

Reviewer: ... and these the CouchDB nodes?

```suggestion
### CouchDB nodes
```

> On each of these nodes run the join command given to you above.
> 
> `docker swarm join --token <very-long-token-value> <main-server-private-ip>:2377`

Reviewer 2: Specificity and code same code format:

```suggestion
On each of these three CouchDB nodes run the join command given to you in step 1 above:

    docker swarm join --token <very-long-token-value> <main-server-private-ip>:2377`

```

> ## Configuration - Node 1 - CHT-core

Reviewer: ```suggestion
## Configuration - CHT Core node
```

> ### Setup Docker

Reviewer: This section is identical to the single node setup right? Either pull this out as its own page and link to it, or use partials to inline a common section, eg:

```
{{< read-content file="some/dir/_partial_cht-core-node-setup.md" >}}
```

Reviewer 2: this and sooooo much more done in t946fb8c

1. `sudo apt update`
2. `sudo apt install docker.io`
3. `sudo apt install docker-compose`

### Directory Structure

Create the following directory structure. This is important.

```
|--CHT
    |__ compose/
    |__ certs/
    |__ upgrade-service/
```

Explanations:

* compose - contains the docker-compose files for cht-core and couchdb.
* certs - the path that you'll put SSL certificates.
* upgrade-service - the path for the docker-compose file of the upgrade-service.

### Download required docker-compose files

The following commands download the 4.0.1 version. If you want a different version, you would have to change `medic:4.0.1` to the version you'd like to install.

> 
> 1. Download the cht-core docker-compose file. Run the following command from the `CHT` directory above.
> `curl -s -o ./compose/cht-core.yml https://staging.dev.medicmobile.org/_couch/builds_4/medic:medic:4.0.1/docker-compose/cht-core.yml`
> 2. Download the upgrade service docker-compose file.
> `curl -s -o ./upgrade-service/docker-compose.yml https://raw.githubusercontent.com/medic/cht-upgrade-service/main/docker-compose.yml`
> 

Reviewer 2: Code formatting 

```suggestion
1. Download the cht-core docker-compose file. Run the following command from the `CHT` directory above:

        curl -s -o ./compose/cht-core.yml https://staging.dev.medicmobile.org/_couch/builds_4/medic:medic:4.0.1/docker-compose/cht-core.yml

2. Download the upgrade service docker-compose file.

        curl -s -o ./upgrade-service/docker-compose.yml https://raw.githubusercontent.com/medic/cht-upgrade-service/main/docker-compose.yml

```

### Docker-compose modifications

> 1. Go to the networks section and add the following:
> 
> ```
>   cht-overlay:
>     driver: overlay
>     external: true
> ```
> 
> 2. Go to each of the containers and add the following:
> 
> ```
>     networks:
>       - cht-overlay
> ```

Reviewer 2: fix indentation (btw - GH isn't letting me put triple ticks  ```  in side suggestions so I'm using indentations :frowning_face:  ) and clarify which files we're editing

Edit both the `cht-core.yml` and `docker-compose.yml` compose files:

```suggestion
1. At the end of each compose file, ensure the 2 `networks:` sections (1 in `cht-core.yml` and 1 in `docker-compose.yml`) look like this by adding three lines:

        networks:
            cht-net:
               name: ${CHT_NETWORK:-cht-net}
            cht-overlay:
               driver: overlay
               external: true

2. For all 6 services in the compose files (5 in `cht-core.yml` and 1 in `docker-compose.yml`), update the `networks:` section to look like this:

        networks:
          - cht-overlay
          - cht-net

```

### Prepare Environment Variables file

> Prepare a `.env` file that contains the following variables and save it in the upgrade-service directory. Note that some of the values are not necessary in this node but for the purpose of making things simple here, we'll create one env file and use it on all nodes.  We are assuming that your home directory is `/home/ubuntu/`. If that's not accurate, please change that in the values below

Reviewer 2: Call out needed actions, light formatting:

```suggestion
Prepare a `.env` file that contains the following variables and save it in the upgrade-service directory. Note that some of the values are not necessary in this node, but for the purpose of making things simple here, we'll create one `.env` file and use it on all nodes.  

We are assuming that your home directory is `/home/ubuntu/`. If that's not accurate, please change that in the values below. 

Be sure to replace these values with real ones:
* `<some-super-long-combination-of-alphanumeric-characters>` 
* `<another-super-long-alphanumeric-characters>`
* `<put-in-a-guid-value>`
```

```
CHT_COMPOSE_PROJECT_NAME=cht
COUCHDB_SECRET=<some-super-long-combination-of-alphanumeric-characters>
DOCKER_CONFIG_PATH=/home/ubuntu/cht/upgrade-service
COUCHDB_DATA=/home/ubuntu/cht/couchdb
CHT_COMPOSE_PATH=/home/ubuntu/cht/compose
COUCHDB_USER=medic
COUCHDB_PASSWORD=<another-super-long-alphanumeric-characters>
CERTIFICATE_MODE=OWN_CERT
COUCHDB_UUID=<put-in-a-guid-value>
SSL_VOLUME_MOUNT_PATH=/etc/nginx/private/
SSL_CERT_FILE_PATH=/etc/nginx/private/cert.pem

> SSL_KEY_FILE_PATH=/etc/nginx/private/key.pem

Reviewer 2: These are the same as the default values in the compose file so we can remove them to simplify the `.env` file

Reviewer 2: done in 946fb8c

```

> **Note:**
> * Make sure that all the directory paths are absolute paths and not relative paths.
> * COUCHDB_UUID can be generated [here](https://www.uuidgenerator.net/)
> * Make sure certificate is named `cert.pem` and key is named `key.pem` - this is right now a requirement because of a known [bug](https://github.com/medic/cht-core/issues/7949)

Reviewer 2: Use native note format:

```suggestion
{{% alert title="Note" %}}
* Make sure that all the directory paths are absolute paths and not relative paths.
* COUCHDB_UUID can be generated [here](https://www.uuidgenerator.net/)
* Make sure certificate is named `cert.pem` and key is named `key.pem` - this is right now a requirement because of a known [bug](https://github.com/medic/cht-core/issues/7949)
{{% /alert %}}
```

### Get the certificate loaded into a cht volume

> Follow the set of steps outlined [here](https://github.com/medic/cht-core/pull/7834#issuecomment-1268710481).

Reviewer 2: let's make this a one stop document and include these instructions inline.

> ## Configuration - Node 2 - couchdb.1

Reviewer: Let's find a way to collapse these three into one so it's clear you're just doing one thing three times.

Reviewer 2: Seconding this!!!

Reviewer 2: done in 946fb8c

### Setup Docker

1. `sudo apt update`
2. `sudo apt install docker.io`
3. `sudo apt install docker-compose`

Create a `cht` directory under your home directory and download files on there.

> 1. Download the clustered cht-couchdb docker-compose file.
> 
> `curl -s -o ./cht/cht-couchdb-clustered.couchdb1.yml https://staging.dev.medicmobile.org/_couch/builds_4/medic:medic:4.0.1/docker-compose/cht-couchdb-clustered.yml`
> 

Reviewer 2: ```suggestion
1. Download the clustered cht-couchdb docker-compose file.

        curl -s -o ./cht/cht-couchdb-clustered.couchdb1.yml https://staging.dev.medicmobile.org/_couch/builds_4/medic:medic:4.0.1/docker-compose/cht-couchdb-clustered.yml

```

```
    networks:
      - cht-overlay
```

## Configuration - Node 4 - couchdb.3

### Setup Docker

1. `sudo apt update`
2. `sudo apt install docker.io`
3. `sudo apt install docker-compose`

Create a `cht` directory under your home directory and download files on there.

1. Download the clustered cht-couchdb docker-compose file.

`curl -s -o ./cht/cht-couchdb-clustered.couchdb3.yml https://staging.dev.medicmobile.org/_couch/builds_4/medic:medic:4.0.1/docker-compose/cht-couchdb-clustered.yml`

2. Create a `.env` file and populate it with the same content above as node 1.

### Docker-compose modifications

1. Since we want to run only one couchdb node per instance, delete couchdb.1 and couchdb.2 containers from the yml.
2. Add this property under the `image` section: `container_name: couchdb.3`
3. Go to the networks section and add the following:

```
  cht-overlay:
    driver: overlay
    external: true
```

4. Go to each of the containers and add the following:

```
    networks:
      - cht-overlay
```

## Launch containers in the different nodes

Please start the containers in the same order as they have been specified here.

> ### Launch `couchdb.2`

Reviewer: I think these names are changing to be FQDN. To save us updating the docs let's just say..

```suggestion
### Launch CouchDB node 2
```
